[RFC] Named macro capture groups (#3649)
rust-lang/rfcs: text/3649-macros-named-capture-groups.md

- Repetition-related diagnostics are limited because the compiler has limited
  ability to guess what a capture group _should_ refer to when a captured
  groups and variables do not align correctly.
- Repetition mismatch diagnostics like can only be emitted after the macro is
  instantiated, rather than when the macro is written. (E.g. "meta-variable
  `foo` repeats 2 times, but `bar` repeats 1 time")
- As a result of the above, using repetition is somewhat fragile; small
  adjustments can break working patterns with little indication of what exactly
  is wrong. Reading code with multiple capture groups can also be confusing.

It is expected that named capture groups will provide a way to remove ambiguity
in expansion and metavariable expressions, as well as unblock diagnostics that
do a better job of guiding the macro mental model.

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

Capture groups can now take a name by providing an identifier and `:` between
the `$` and opening `(`. This group can then be referred to by name in the
expansion:

```rust
macro_rules! foo {
    ( $group1:( $a:ident ),+ ) => {
        $group1( println!("{}", $a); )+

>     }

Reviewer: when you write this I thought you are defining a metavariable `$group1` which matches a comma-separated list of `$a:ident`, and using `$group1` will expand to `foo,bar,baz,etc`.

but no the `$group1` is just a label :thinking: 

the rationale mentioned using `:` is to be "to be more consistent with existing fragment specifiers" but IMO similarity with fragment specifiers  *is exactly why this RFC should not use the current syntax*.

i prefer getting rid of `:` (my main source of confusion) and use a label syntax like

```rust
macro_rules! make_functions {
    ( 
        names: [ $'names $($name:ident),+ ],
//               ^~~~~~~~~
        $'greetings $( greeting: $greeting:literal, )?
    ) => {
        $'names $(
            fn $name() {
                println!("function {} called", stringify!($name));
                $'greetings $(println!("{}", $greeting) )?
            }
        )+
    }
}

macro_rules! innermost1 {
    ( $'outer_rep $($a:ident: $'inner_rep $($b:literal),* );+ ) => {
        [$'outer_rep $( $'inner_rep $( ${index($'outer_rep)}, )* )+]
    };
}
```

Reviewer 2: Metavariables matching some parts of the input are an interesting idea in general.
If we also introduce "match exactly once repetitions", then we'll also be able to capture just arbitrary token streams.
It would be nice to also be able to bind a name to "repeated exactly one" groups aka just artibtrary substreams in the input.
```rust
// macro LHS, capture "exactly once" repetition
$my_tokens:(a b $var c d)①

// macro RHS, reemit (although finding a good way to substitute variables is an interesting problem)
$my_tokens(value_for_var)
```

Author: Thanks for the ideas - I updated the default syntax to be `$group1(...)` (no colon) and added the label syntax as a possibility. 

>  It would be nice to also be able to bind a name to "repeated exactly one" groups aka just artibtrary substreams in the input.

Is there a standard kleene for a single capture, or could we get by with no kleene at all? I could probably include a  proposal for single captures in this RFC, to be accepted as part of it or rejected separately.

Reviewer 3: In regexes, a "capture exactly once" group is just `(...)`.

Author: After writing some more examples, it seems like "capture once" and "emit the entire capture" would be pretty useful, so I just included them as part of the proposal. Especially for code that just needs to be validated and passed elsewhere, or matching and reemitting optional keywords/tts like `mut`, `&` or `pub(crate)` (just allowing capture groups that don't contain metavars is enough for that, but `$pub_crate` in the expansion is nicer to read than `$pub_crate(pub(crate))`).

Reviewer 2: One advantage of the reemitting is that it will use spans from the input, instead of spans from the macro body.

One example I recently observed in https://github.com/rust-lang/rust/pull/119412 is the `mir!` macro in standard library.
It will match on something like `let $expr = $expr;` and then reemit `$expr = $expr;` as an assignment expression.
If `=` is emitted manually from the macro it will get the macro span and may mess up the combined span of the assignment expression.
If `=` is taken from the input, then the expression will get a precise span from the input, which is good for any kind of DSLs.

in place of a variable and an index, since these three expressions relate more
to how entire _groups_ are expanded than the variables they take as arguments.

Further reading:

- [`macro_metavar_expr` RFC][`macro_metavar_expr`] and
  [tracking issue](https://github.com/rust-lang/rust/issues/83527)
- [Proposal for possible specific behavior](https://github.com/rust-lang/rust/pull/122808#issuecomment-2124471027)

# Drawbacks
[drawbacks]: #drawbacks

Why should we *not* do this?

- If [`macro_metavar_expr`] stabilizes before this merges, this will add a
  duplicate way of using those expressions. If this RFC is accepted,
  stabilizing only a subset of metavariable expressions that does not conflict
  should be considered.

# Rationale and alternatives
[rationale-and-alternatives]: #rationale-and-alternatives

- Variable definition syntax could be `$ident(/* ... */)` rather than
  `$ident:(/* ... */)`. Including the `:` is proposed to be more consistent
  with existing fragment specifiers.
- There is room for macros to become smarter in their expansions without adding
  named capture groups. As mentioned elsewhere in this RFC, it seems like

>   adding named groups is a cleaner solution with less cognifitive

Reviewer 4: ```suggestion
  adding named groups is a cleaner solution with less cognitive overhead.
```

                // If a greeting is provided, print it in every function
                $greetings( println!("{}", $greeting) )?
            }
        )+
    }
}

fn main() {
    make_functions! {
        names: [foo, bar],
        greeting: "hello!",
    }

    foo();
    bar();

    // output:
    // function foo called
    // hello!
    // function bar called
    // hello!
}

```

This expansion is not easily possible without named capture groups because
of ambiguity regarding which groups are referred to.

Expansion of the above will approximately follow this procedural model:

```rust
let mut ret = TokenStream::new();

// Append an expansion for each time group1 is matched
for NamesCaptures { name } in greetings {
    let mut fn_body = quote! { println!("function {} called", stringify!($name)); };

    // Append the greeting for each 
    for GreetingCaptures { greeting } in greetings {
        fn_body += quote! { println!("{}", #greeting) };
    }

    // Construct the function item and append to returned tokens
    ret += quote! { fn #name() { #fn_body  } };
}
```

Groups can also be used in the expansion without `(...)` to emit their entire

> capture.

Reviewer 5: I love this specific point, and have wanted this in the past.

// - `outer_a` repeats three times within root, so repeats 3x within `outer_b`
// - `inner` repeats `[2, 0, 0]` times within the `outer_a` instances. This
//   drives how often `x` and `y` get repeated within that group.
$outer_b( ob[$outer_a( oa[$y $middle( $inner( i[$x $y] ) )] )] )

// Forbidden: `x` is missing parent `outer_a`
$outer_b( $middle( $inner( $x $y ) ) ) )

// Forbidden: group nesting within itself
$outer_b( $outer_b( $y ) )
$outer_b( $outer_b )
```

#### Single matches

Capture groups must currently specify a kleene operator (`*`, `+`, or `?`) that
determines if the group should match zero or more times, one or more times, or
up to one time. This RFC will allow omitting the kleene operator to indicate
that a group must be captured exactly once. That is, `$group(foo)` is a valid
pattern (with current rules, `$(foo)` is forbidden).

With this "exactly once" match there is no purpose in having a repetition token
(e.g. the comma in `$(...),*`), so it must be omitted.

#### Entire group expansion

Since groups are named, it is now possible to write a group name to reemit its
captured contents with no further expansion. This syntax uses the group name

> but omits the `(/* expansion pattern */)/* kleene */`:

Reviewer 5: One potential source of ambiguity:

How would you express "I want to expand the entire named group, followed by some tokens in parentheses"?

I have a suggestion for how to solve that: `${$group}(tokens)`.

tokens and anything that would be bound to a metavariable. Span from the
macro invocation can be kept here, which should improve the diagnosability of
some macros.

The above rules regarding allowed group usage locations must still be followed.

### Metavariable Expressions

This RFC proposes some changes to metavariable expressions that will leverage
named groups to hopefully make them more user-friendly.

_At time of writing, part of macro metavariable expressions are under
consideration for stabilization. Depending on what is selected, these rules may
need to change slightly._

- `${index($group)}`: Return the number of times that the group has been
  _expanded_ so far. Must be used within the group that is given as an
  argument.
- `${count($metavar)}`: Return the number of times a group or metavariable was
  _captured_.
- `${len($metavar)}`: Because `count` becomes more flexible, `len` is no longer
  needed and can be removed.
- `${ignore($metavar)}`: if this RFC is accepted than `$ignore` can be removed.
  It is used to specify which capture group an expansion group belongs to when
  no metavariables are used in the expansion; with named groups, however, this

>   is specified by the group name rather than by contained metavariables.

Reviewer 5: :tada: This is a great sign that this design is a good idea.

// Ok: prints `[2 0 0]`
[$outer_a( ${count($inner} )*]

// Ok: prints `[2 0 0]`
[$outer_a( ${count($x} )*]

// Ok: prints `o[m[x 2]] o[] o[m[x 0] m[x 0]]`
// `$middle[0,0]``"sees" two `$x` captures. `$middle[2,0]` and `$middle[2,1]` both
// don't see any.
$outer_a( o[$middle( m[x ${count($x)}] )*] )*

/* looking at ancestors */

// Ok: prints [3 3 3]
// `count` is used at level 2 (one deeper than `outer_a`), so it sees all
// `outer_a`s each time.
[$outer_a( ${count($outer_a)} )*]

// Ok: prints [[3] [] [3 3]]
// Similar to the above
[$outer_a( [$middle( ${count($outer_a)} )*] )*]

/* errors */

// Error: trying to count a variable that is neither a descendent or anscestor
[$outer_a( ${count($y} )*]
```

TODO: we could relax the final rule and allow counting siblings of ancestors

# Drawbacks
[drawbacks]: #drawbacks

Why should we *not* do this?

- If [`macro_metavar_expr`] stabilizes before this merges, this will add a
  duplicate way of using those expressions. If this RFC is accepted,
  stabilizing only a subset of metavariable expressions that does not conflict
  should be considered.

# Rationale and alternatives
[rationale-and-alternatives]: #rationale-and-alternatives

> 

Reviewer 5: ```suggestion

- In macro metavariable syntax, using named capture groups, we could treat `count` and `index` as *fields* rather than *functions*. For instance, we could write `${$group.index}` rather than `${index($group)}`. This would be consistent with the [macro fragment fields](https://github.com/rust-lang/rfcs/pull/3714) proposal.
- Since we no longer need both `count` and `len`, we could choose to use either name for the remaining function we still provide. We should consider whether `count` or `len` best describes this functionality.
```

> - Variable definition syntax could be `$ident(/* ... */)` rather than
>   `$ident(/* ... */)`. Including the `:` is proposed to be more consistent
>   with existing fragment specifiers.

Reviewer 6: Boths sides of this `rather than` are identical.
deab736c135f82a346544682eaa4a9a90627512d swapped the proposed and alternative syntax and removed the `:` on the right hand side without adding it on the left.

```suggestion
- Variable definition syntax could be `$ident:(/* ... */)` rather than
  `$ident(/* ... */)`. Including the `:` is proposed to be more consistent
  with existing fragment specifiers.
```

  instantiated, rather than when the macro is written. (E.g. "meta-variable
  `foo` repeats 2 times, but `bar` repeats 1 time")
- As a result of the above, using repetition is somewhat fragile; small
  adjustments can break working patterns with little indication of what exactly
  is wrong. Reading code with multiple capture groups can also be confusing.
- Metavariable expressions as they currently exist use an unintuitive format:
  syntax like `${count($var, n)}` is used to refer to the `n`th ancestor group
  of the smallest group that captures `$var`. Referring to groups directly would
  be more straightforward than using a proxy.

It is expected that named capture groups will provide a way to remove ambiguity
in expansion and metavariable expressions, as well as unblock diagnostics that
do a better job of guiding the macro mental model.

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

Capture groups can now take a name by providing an identifier between the `$`
and opening `(`. This group can then be referred to by name in the expansion:

```rust
macro_rules! foo {
    ( $group1( $a:ident ),+ ) => {
        $group1( println!("{}", $a); )+
    }
}

> ```

Reviewer 7: As mentioned by other reviewers, the expansion-side macro is already valid.

I'd like to suggest an alternative:

```rust
macro_rules! foo {
    ( $group1( $a:ident ),+ ) => {

        ${for a in group1} (
            println!("{}", $a);
        )
    }
}
```

This alternative is more verbose, but it's also much more explicit about what's going on.

```rust
let mut ret = TokenStream::new();

// Append an expansion for each time group1 is matched
for NamesCaptures { name } in greetings {
    let mut fn_body = quote! { println!("function {} called", stringify!($name)); };

    // Append the greeting for each 
    for GreetingCaptures { greeting } in greetings {
        fn_body += quote! { println!("{}", #greeting) };
    }

    // Construct the function item and append to returned tokens
    ret += quote! { fn #name() { #fn_body  } };
}
```

Groups can also be used in the expansion without `(...)` to emit their entire
capture.

This works well with a new "match exactly once" grouping that takes no kleene
operator (as opposed to matching zero or more times (`*`), matching once or
more (`+`), or matching zero or one times (`?`)).

```rust
macro_rules! rename_fn {
    (
        $newname:ident;
        $pub(pub)? fn $oldname:ident( $args( $mut(mut)? $arg:ident: $ty:ty )* );
    ) => {
        $pub fn $newname( $args );
    }
}
```

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

Macro captures currently include the following grammar node:

> > `$` ( _MacroMatch<sup>+</sup>_ ) _MacroRepSep_<sup>?</sup> _MacroRepOp_

Reviewer 8: ```suggestion
> `$` `(` _MacroMatch<sup>+</sup>_ `)` _MacroRepSep_<sup>?</sup> _MacroRepOp_
```

This will be expanded to the following:

> > `$` ( IDENTIFIER_OR_KEYWORD except crate | RAW_IDENTIFIER | _ )<sup>?</sup> ( _MacroMatch<sup>+</sup>_ ) _MacroRepSep_<sup>?</sup> _MacroRepOp_

Reviewer 8: I commented elsewhere about concerns with the proposed syntax for exactly-once groups. But with the current proposal, I think the grammar node would need the trailing MacroRepSep-MacroRepOp sequence to be optional.

```suggestion
> `$` ( IDENTIFIER_OR_KEYWORD except crate | RAW_IDENTIFIER | _ )<sup>?</sup> `(` _MacroMatch<sup>+</sup>_ `)` ( _MacroRepSep_<sup>?</sup> _MacroRepOp_ )<sup>?</sup>
```

(I also added missing backticks around the literal parentheses tokens that delimit the group body.)

As a result, `$identifier( /* ... */ ) /* sep and kleene */` will allow naming
a capture group. It can then be used in expansion:

```rust
$identifier(
  /* expansion within group */
) /* sep and kleene */
```

Group and metavariables share the same namespace; all groups and metavariables

> must have a unique name within the macro capture pattern.

Reviewer 9: Syntax like e.g. `$identifier( /* … */ ),` is *already* valid and useful in macro expansions.
```rs
macro_rules! call_many_fuctions {
    ([$($($function_name:ident),+$(,)?)?]($arg:expr)) => {
        [
            $($(
                $function_name($arg),
            )+)?
        ]
    }
}
fn foo(x: i32) -> i32 { x }
fn bar(x: i32) -> i32 { x + 1 }
fn baz(x: i32) -> i32 { x * x }
fn qux(x: i32) -> i32 { x / 2 }

fn main() {
    println!("{:?}", call_many_fuctions!([foo, bar, baz, qux](100)));
}
```
(this concrete example doesn’t follow it up with a `*`, `+` or `?`, but these aren’t forbidden either, of course)

If this RFC proposes to make the “parsing” of the macro RHS depend on the “type” of the fragment, I think that’s a clear possible disadvantage.

Especially with the (IMO very *useful*) idea that `$group` on its own can already stand in for an expansion of the whole group’s contents, so expanding some identifier `$function_name` – or expression, pattern, etc… – doesn’t work different from expanding some group `$group` – it seems desirable (to me) if `$function_name($arg), *etc…` doesn’t have a severely different meaning from `$group($arg), *etc…` either.

Reviewer: the design that `$group` does not expand is primarily why i suggested in https://github.com/rust-lang/rfcs/pull/3649#discussion_r1618564913 that it should use a separate syntax `$'group`.

Author: These are great points. I wish we could use the same syntax for metavars and groups, but the `$'group` syntax does get us out of that, so I'll use that for the implementation and leave an unresolved question.

containing only `$a`).

Rewritten to use named groups instead:

```rust
macro_rules! innermost1 {
    ( $outer_rep( $a:ident: $inner_rep( $b:literal ),* );+ ) => {
        [$outer_rep( $inner_rep( ${index($outer_rep)}, )* )+]
    };
}
```

It is significantly easier to see what the call to `index` is referring to. As
an added benefit, its meaning will not change if its position is moved in the
code (e.g. moving to be within `$outer_rep`, but not `$inner_rep`).

This RFC proposes that `count`, `index`, and `len` will accept group names
in place of a variable and an index, since these three expressions relate more
to how entire _groups_ are expanded than the variables they take as arguments.

Further reading:

- [`macro_metavar_expr` RFC][`macro_metavar_expr`] and
  [tracking issue](https://github.com/rust-lang/rust/issues/83527)
- [Proposal for possible specific behavior](https://github.com/rust-lang/rust/pull/122808#issuecomment-2124471027)

### "Exactly one" matching and full group emission

If a group is specified without a kleene operator (`*`, `+`, `?`), it will now
be assumed to match exactly once. This will be most useful with the ability to

> emit an entire matched group.

Reviewer 8: If exactly-once groups don't have a kleene operator to terminate the group specifier, how would a macro author define a rule that matches an exactly-once group followed by one of the tokens `*`, `+`, or `?`?

For example, in the following (admittedly kind of silly) example, I want to specify a multiplication-like syntax for my macro. But I think the `*` would be attached to the group instead, silently turning it into a zero-or-more group.

```rust
macro_rules! example1 {
    ( $g1($a:ident[$b:expr]) * $g2($c:ident[$d:expr]) ) => { $g1 $g2 }
}

example1!(x[1] * y[2]);  // ERROR: Something something unexpected `*`...
```

Further, a kleene-looking token could &ldquo;reach back&rdquo; across an intermediate token, interpreting it as a separator. I think this would be extra surprising, because a user may edit a macro definition and without thinking change the the nature of the group at a distance.

For example, the following example would specify an exactly-once group.

```rust
macro_rules! example2 {
    ( $g($a:ident $b:ident) x y + ) => { $g }
}

example2!(foo bar x y +);  // Ok
```

But if I later edited it to remove the `y` token, it would silently become a one-or-more group with `x` as the separator.

```rust
macro_rules! example2 {
    ( $g($a:ident $b:ident) x + ) => { $g }
}

example2!(foo bar x +);  // Error: Something about `+` is not an ident or unexpected end of input
```

For this reason, I don't think it would be a satisfying solution to just add an escape syntax for `*`, `+`, and `?`. It would _work_, but it would still lead to gotchas like this where the programmer has to reason out why their macro silently broke before eventually slapping their head and adding an escape.

I think it would be better to introduce a new &ldquo;kleene&rdquo; operator for this new group type. (Scare quotes since Kleene didn't actually define a regular language operator like that, but he wasn't concerned with matches anyway.) There isn't any symbol that I *love* for the job&mdash;maybe `!` or `=`?

Reviewer 8: I think there would be some backwards compatibility concerns with my suggestion if we allowed unnamed group matchers to use the new exactly-once operator. For example, consider the following macro definition.

```rust
macro_rules! example {
    ( $( $a:ident )!* ) => { $( $a ) }
}
```

Today, the rule matches zero or more idents with `!` as the separator. If we made `!` the new exactly-once operator, then I think this rule would instead match exactly one ident followed by `*`.

Maybe we only allow exactly-once groups with the new named syntax?

captured contents with no further expansion. This syntax uses the group name
but omits the `(/* expansion pattern */)/* kleene */`:

```rust
// Ok: prints `ob[y0]`
$outer_b
```

The entire contents of the capture group are emitted, including both exact
tokens and anything that would be bound to a metavariable. Span from the
macro invocation can be kept here, which should improve the diagnosability of
some macros.

The above rules regarding allowed group usage locations must still be followed.

### Metavariable Expressions

This RFC proposes some changes to metavariable expressions that will leverage
named groups to hopefully make them more user-friendly.

_At time of writing, part of macro metavariable expressions are under
consideration for stabilization. Depending on what is selected, these rules may
need to change slightly._

- `${index($group)}`: Return the number of times that the group has been
  _expanded_ so far. Must be used within the group that is given as an
  argument.
- `${count($metavar)}`: Return the number of times a group or metavariable was
  _captured_.
- `${len($metavar)}`: Because `count` becomes more flexible, `len` is no longer
  needed and can be removed.

> - `${ignore($metavar)}`: if this RFC is accepted than `$ignore` can be removed.

Reviewer 5: ```suggestion
- `${ignore($metavar)}`: if this RFC is accepted than `ignore` can be removed.
```

Reviewer 10: r/than/then

    }

    foo();
    bar();

    // output:
    // function foo called
    // hello!
    // function bar called
    // hello!
}

```

This expansion is not easily possible without named capture groups because
of ambiguity regarding which groups are referred to.

Expansion of the above will approximately follow this procedural model:

```rust
let mut ret = TokenStream::new();

// Append an expansion for each time group1 is matched
for NamesCaptures { name } in greetings {
    let mut fn_body = quote! { println!("function {} called", stringify!($name)); };

    // Append the greeting for each
    for GreetingCaptures { greeting } in greetings {
        fn_body += quote! { println!("{}", #greeting) };
    }

    // Construct the function item and append to returned tokens
    ret += quote! { fn #name() { #fn_body  } };
}
```

Groups can also be used in the expansion without `(...)` to emit their entire
capture.

This works well with a new "match exactly once" grouping that takes no kleene
operator (as opposed to matching zero or more times (`*`), matching once or
more (`+`), or matching zero or one times (`?`)).

```rust
macro_rules! rename_fn {
    (
        $newname:ident;

>         $pub(pub)? fn $oldname:ident( $args( $mut(mut)? $arg:ident: $ty:ty )* );

Reviewer 11: When reading, I felt the args repetition was missing a separator, e.g. `,`?

```suggestion
        $pub(pub)? fn $oldname:ident( $args( $mut(mut)? $arg:ident: $ty:ty ),* );
```

- Variable definition syntax could be `$ident:(/* ... */)` rather than
  `$ident(/* ... */)`. Including the `:` is proposed to be more consistent
  with existing fragment specifiers.
- There is room for macros to become smarter in their expansions without adding
  named capture groups. As mentioned elsewhere in this RFC, it seems like
  adding named groups is a cleaner solution with less cognitive overhead.

# Prior art
[prior-art]: #prior-art

- Regex allows the naming of reepeating capture groups for expansion, including
  those that do not capture anything else.

# Unresolved questions
[unresolved-questions]: #unresolved-questions

- Syntax: the original proposal was to include a colon, e.g.
  `$group1:(/* ... */)`. A label-like syntax of `$'group1 $(/* ... */)` was
  also proposed.

**Possibly edition-sensitive** the proposed syntaxes are currently rejected
under the `missing_fragment_specifier` lint. That means that
`#![allow(missing_fragment_specifier)]` makes rustc accept the proposed syntax
as valid, which could conflict with this proposal.

# Future possibilities

> [future-possibilities]: #future-possibilities

Reviewer 4: I meant same number of repetitions, not tokens. so you can have `$foo(a a) $foo(b)` match `a a a a b b`
